File: python/src/solana_mpp/channel_store.py
"""Richer channel store for the session intent.

Unlike the charge replay :class:`~solana_mpp.store.Store` (a flat consumed-key
set), the session server tracks live channel lifecycle state and advances a
settled watermark under concurrent vouchers. This needs an atomic
read-modify-write store, mirroring the Rust spine ``ChannelStore`` /
``MemoryChannelStore`` at ``rust/crates/mpp/src/store.rs``.

The ``update_channel`` method runs the caller's updater closure inside the
store lock so the entire read-modify-write is atomic and no concurrent voucher
can interleave a double-spend.
"""

from __future__ import annotations

import asyncio
from collections.abc import Callable
from dataclasses import dataclass, field
from typing import Any, Protocol, runtime_checkable


@dataclass
class PendingDelivery:
    """A delivery reserved by the server but not yet committed by the client."""

    delivery_id: str
    amount: int
    sequence: int
    expires_at: int

    def to_dict(self) -> dict[str, Any]:
        return {
            "deliveryId": self.delivery_id,
            "amount": self.amount,
            "sequence": self.sequence,
            "expiresAt": self.expires_at,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> PendingDelivery:
        return cls(
            delivery_id=data["deliveryId"],
            amount=int(data["amount"]),
            sequence=int(data["sequence"]),
            expires_at=int(data["expiresAt"]),
        )


@dataclass
class CommittedDelivery:
    """A committed delivery, kept for idempotent commit replay."""

    delivery_id: str
    amount: int
    cumulative: int
    voucher_signature: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "deliveryId": self.delivery_id,
            "amount": self.amount,
            "cumulative": self.cumulative,
            "voucherSignature": self.voucher_signature,
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> CommittedDelivery:
        return cls(
            delivery_id=data["deliveryId"],
            amount=int(data["amount"]),
            cumulative=int(data["cumulative"]),
            voucher_signature=data["voucherSignature"],
        )


@dataclass
class ChannelState:
    """Persisted state of a payment channel, managed by the server."""

    channel_id: str
    authorized_signer: str
    deposit: int
    cumulative: int = 0
    finalized: bool = False
    highest_voucher_signature: str | None = None
    highest_voucher_expires_at: int | None = None
    close_requested_at: int | None = None
    operator: str | None = None
    next_delivery_sequence: int = 0
    pending_deliveries: list[PendingDelivery] = field(default_factory=list)
    committed_deliveries: list[CommittedDelivery] = field(default_factory=list)

    def clone(self) -> ChannelState:
        """Return a deep copy so updater closures cannot mutate stored state."""
        return ChannelState(
            channel_id=self.channel_id,
            authorized_signer=self.authorized_signer,
            deposit=self.deposit,
            cumulative=self.cumulative,
            finalized=self.finalized,
            highest_voucher_signature=self.highest_voucher_signature,
            highest_voucher_expires_at=self.highest_voucher_expires_at,
            close_requested_at=self.close_requested_at,
            operator=self.operator,
            next_delivery_sequence=self.next_delivery_sequence,
            pending_deliveries=[
                PendingDelivery(d.delivery_id, d.amount, d.sequence, d.expires_at) for d in self.pending_deliveries
            ],
            committed_deliveries=[
                CommittedDelivery(d.delivery_id, d.amount, d.cumulative, d.voucher_signature)
                for d in self.committed_deliveries
            ],
        )


@runtime_checkable
class ChannelStore(Protocol):
    """Async store for channel state with atomic watermark advancement.

    Implementations MUST guarantee that ``update_channel`` is atomic to
    prevent double-spend under concurrent requests.
    """

    async def get_channel(self, channel_id: str) -> ChannelState | None: ...

    async def put_channel(self, channel_id: str, state: ChannelState) -> None: ...

    async def update_channel(
        self,
        channel_id: str,
        updater: Callable[[ChannelState | None], ChannelState],
    ) -> ChannelState: ...

    async def mark_finalized(self, channel_id: str) -> None: ...


class MemoryChannelStore:
    """In-memory channel store backed by an asyncio lock.

    The ``update_channel`` method holds the lock across the read, the updater
    call, and the write, so the entire read-modify-write is atomic.
    """

    def __init__(self) -> None:
        self._data: dict[str, ChannelState] = {}
        self._lock = asyncio.Lock()

    async def get_channel(self, channel_id: str) -> ChannelState | None:
        state = self._data.get(channel_id)
        return state.clone() if state is not None else None

    async def put_channel(self, channel_id: str, state: ChannelState) -> None:
        async with self._lock:
            self._data[channel_id] = state.clone()

    async def update_channel(
        self,
        channel_id: str,
        updater: Callable[[ChannelState | None], ChannelState],
    ) -> ChannelState:
        async with self._lock:
            current = self._data.get(channel_id)
            # Hand the updater a clone so a raising updater cannot leave the
            # stored state partially mutated.
            new_state = updater(current.clone() if current is not None else None)
            self._data[channel_id] = new_state.clone()
            return new_state.clone()

    async def mark_finalized(self, channel_id: str) -> None:
        async with self._lock:
            state = self._data.get(channel_id)
            if state is None:
                raise KeyError(f"Channel {channel_id} not found")
            state.finalized = True

File: ruby/lib/mpp/channel_store.rb
# frozen_string_literal: true

module Mpp
  # Persisted state of a payment channel, managed by the session server.
  #
  # Unlike the charge replay store (a write-once consumed-signature set), the
  # session server tracks lifecycle: the settled watermark, the authorized
  # signer, pending/committed metered deliveries, and close/finalize flags.
  #
  # Mirrors `rust/crates/mpp/src/store.rs` (ChannelState / ChannelStore).
  class ChannelState
    attr_accessor :channel_id, :authorized_signer, :deposit, :cumulative,
      :finalized, :highest_voucher_signature, :highest_voucher_expires_at,
      :close_requested_at, :operator, :next_delivery_sequence,
      :pending_deliveries, :committed_deliveries

    def initialize(channel_id:, authorized_signer:, deposit:, cumulative: 0,
      finalized: false, highest_voucher_signature: nil,
      highest_voucher_expires_at: nil, close_requested_at: nil, operator: nil,
      next_delivery_sequence: 0, pending_deliveries: [], committed_deliveries: [])
      @channel_id = channel_id
      @authorized_signer = authorized_signer
      @deposit = deposit
      @cumulative = cumulative
      @finalized = finalized
      @highest_voucher_signature = highest_voucher_signature
      @highest_voucher_expires_at = highest_voucher_expires_at
      @close_requested_at = close_requested_at
      @operator = operator
      @next_delivery_sequence = next_delivery_sequence
      @pending_deliveries = pending_deliveries
      @committed_deliveries = committed_deliveries
    end

    # Deep copy so a store updater never mutates the persisted object in place.
    def dup_state
      ChannelState.new(
        channel_id: channel_id,
        authorized_signer: authorized_signer,
        deposit: deposit,
        cumulative: cumulative,
        finalized: finalized,
        highest_voucher_signature: highest_voucher_signature,
        highest_voucher_expires_at: highest_voucher_expires_at,
        close_requested_at: close_requested_at,
        operator: operator,
        next_delivery_sequence: next_delivery_sequence,
        pending_deliveries: pending_deliveries.map(&:dup),
        committed_deliveries: committed_deliveries.map(&:dup)
      )
    end
  end

  # A delivery reserved by the server but not yet committed by the client.
  class PendingDelivery
    attr_accessor :delivery_id, :amount, :sequence, :expires_at

    def initialize(delivery_id:, amount:, sequence:, expires_at:)
      @delivery_id = delivery_id
      @amount = amount
      @sequence = sequence
      @expires_at = expires_at
    end

    def dup
      PendingDelivery.new(delivery_id: delivery_id, amount: amount, sequence: sequence, expires_at: expires_at)
    end
  end

  # A committed delivery, kept for idempotent commit replay.
  class CommittedDelivery
    attr_accessor :delivery_id, :amount, :cumulative, :voucher_signature

    def initialize(delivery_id:, amount:, cumulative:, voucher_signature:)
      @delivery_id = delivery_id
      @amount = amount
      @cumulative = cumulative
      @voucher_signature = voucher_signature
    end

    def dup
      CommittedDelivery.new(delivery_id: delivery_id, amount: amount, cumulative: cumulative, voucher_signature: voucher_signature)
    end
  end

  # Raised when a store update closure rejects a transition.
  class StoreError < StandardError; end

  # Channel store interface with atomic read-modify-write.
  #
  # Implementations MUST guarantee that `update_channel` is atomic so no
  # concurrent voucher/commit/topup/close can interleave and double-spend.
  class ChannelStore
    def get_channel(_channel_id)
      raise NotImplementedError
    end

    def put_channel(_channel_id, _state)
      raise NotImplementedError
    end

    # Atomically read-modify-write. The block receives the current state (or
    # nil if absent) and returns the new state, or raises StoreError to abort.
    def update_channel(_channel_id)
      raise NotImplementedError
    end

    def mark_finalized(channel_id)
      update_channel(channel_id) do |state|
        raise StoreError, "Channel #{channel_id} not found" if state.nil?

        state.finalized = true
        state
      end
    end
  end

  # Thread-safe in-memory channel store for tests and single-process servers.
  #
  # A single mutex guards the whole map; `update_channel` holds it across the
  # entire read-modify-write, giving the atomicity the session server relies on.
  class MemoryChannelStore < ChannelStore
    def initialize
      @mutex = Mutex.new
      @channels = {}
    end

    def get_channel(channel_id)
      @mutex.synchronize do
        state = @channels[channel_id]
        state&.dup_state
      end
    end

    def put_channel(channel_id, state)
      @mutex.synchronize { @channels[channel_id] = state.dup_state }
    end

    def update_channel(channel_id)
      @mutex.synchronize do
        current = @channels[channel_id]
        new_state = yield(current&.dup_state)
        raise StoreError, "channel updater returned nil" if new_state.nil?

        @channels[channel_id] = new_state.dup_state
        new_state
      end
    end
  end
end
